### PR TITLE
809 set properties in http request message

### PR DESCRIPTION
**Pull request by nicostouch**
**What kind of change does this PR introduce?**
This PR introduces a new feature to Refit. Being able to pass runtime state into the `HttpRequestMessage.Properties` dictionary in order to access it inside a `DelegatingHandler` by adding a `Property` attribute to a parameter.

    public interface IGitHubApi
    {
        [Post("/users/new")]
        Task CreateUser([Body] User user, [Property("SomeKey")] string someValue);

        [Post("/users/new")]
        Task CreateUser([Body] User user, [Property] string someOtherKey);
    }

    class RequestPropertyHandler : DelegatingHandler
    {
        public RequestPropertyHandler(HttpMessageHandler innerHandler = null) : base(innerHandler ?? new HttpClientHandler()) {}

        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        {
            // See if the request has a the property
            if(request.Properties.ContainsKey("SomeKey")
            {
                var someProperty = request.Properties["SomeKey"];
                //do stuff
            }

            if(request.Properties.ContainsKey("someOtherKey")
            {
                var someOtherProperty = request.Properties["someOtherKey"];
                //do stuff
            }

            return await base.SendAsync(request, cancellationToken).ConfigureAwait(false);
        }
    }

PR also includes:
- full documentation on 
  - how to use the feature
  - behavior of edge cases / internals 
- comprehensive set of unit tests covering:
  - adding a property with a key
  - adding a property without specifying its key
  - adding properties with duplicate keys
  - how the feature interacts with Query, Header, and Multipart parameters

**What is the current behavior?**
I opened issue https://github.com/reactiveui/refit/issues/809 for it about a year ago.

Currently Refit doesn't support doing this natively, and it can only be worked around in a pretty hacky way with strings only.
That issue has seen some interest since opening it. The feature is available in RestEase and on the roadmap for Flurl.

**What might this PR break?**
The excellent code-coverage of the unit tests has ensured it hasn't broken anything fundamentally. 
Additionally I've taken great care to add enough coverage to tease out any subtle bugs the feature might have introduced into Query, Header or Multipart functionality. Happy to add any additional test cases if deemed necessary.

**Please check if the PR fulfills these requirements**
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Other information**:
I love Refit and this closes a functionality gap making Refit the only one out of Refit, RestSharp, RestEase, and Flurl to support `HttpClient`, `IHttpClientFactory`, and populating properties into `HttpRequestMessage.Properties`

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/reactiveui/refit?pullRequest=1001) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/reactiveui/refit?pullRequest=1001) before we can accept your contribution.<br/><br/>:x: nicostouch [sign now](https://cla.dotnetfoundation.org/reactiveui/refit?pullRequest=1001)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/reactiveui/refit?pullRequest=1001) it.</sub>

File: Refit/Attributes.cs [hunk 1]
```diff
@@ -206,6 +206,19 @@ public HeaderAttribute(string header)
         public string Header { get; }
     }
 
+    [AttributeUsage(AttributeTargets.Parameter)]
```
Comment (clairernovotny) [inline Refit/Attributes.cs:209]: This type should have doc comments on it and its members

File: Refit/RequestBuilderImplementation.cs [hunk 2]
```diff
@@ -666,6 +675,11 @@ private async Task<T> DeserializeContentAsync<T>(HttpResponseMessage resp, HttpC
                     }
                 }
 
+                foreach (var property in propertiesToAdd)
+                {
+                    ret.Properties.Add(property.Key, property.Value);
```
Comment (clairernovotny) [inline Refit/RequestBuilderImplementation.cs:680]: I know there's a test for the Dup, but Add won't throw here if the key is already present?

**Review by clairernovotny — CHANGES_REQUESTED** [review]
This PR looks great, just one small thing and I can merge it.

File: Refit/Attributes.cs (hunk 1 above)
Comment (nicostouch) [reply Refit/Attributes.cs:214]: I have added some docs to it :)

File: Refit/RequestBuilderImplementation.cs (hunk 2 above)
Comment (nicostouch) [reply Refit/RequestBuilderImplementation.cs]: I've updated this to use the indexer rather than Add method. In the first instance though propertiesToAdd never contains duplicate keys. Also since Refit itself is constructing the HttpRequestMessage before any DelegatingHandlers get to run it's going to be empty, so I think it's likely OK, but have changed it to more of a 'last write wins' in keeping with how the rest of the feature works and erring on the side of not blowing up if it ever happens.
